src/Concordium/Client/GRPC2.hs

```diff
@@ -2552,20 +2852,234 @@ getAccountInfoV2 ::
-getNextSequenceNumberV2 accAddress = withUnaryCoreV2 (callV2 @"getNextAccountSequenceNumber") msg ((fmap . fmap) fromProto)
+getNextSequenceNumberV2 accAddress = withUnaryV2 (callV2 @"getNextAccountSequenceNumber") msg ((fmap . fmap) fromProto)
   where
     msg = toProto accAddress
 
+-- |Send a request to a unary V2 GRPC API endpoint and return the result.
+withUnaryV2 ::
+    ( HasMethod CS.Queries m,
+      MonadIO n,
+      i ~ MethodInput CS.Queries m,
+      o ~ MethodOutput CS.Queries m ) =>
+    -- |The procedure to invoke.
+    RPC CS.Queries m ->
+    -- |The procedure input.
+    i ->
+    -- |A mapping of the result.
+    (GRPCResult o -> b) ->
+    ClientMonad n b
+withUnaryV2 method input k = withGRPCCoreV2 callHelper k
+    where
+        -- This is here so we may leverage `withGRPCCoreV2`.
+        callHelper client = do
+            res <- rawUnary method client input
+            return $ fmap RawUnaryOutput res
+
+-- |Invoke a streaming V2 GRPC API endpoint and return the (collected) results in a sequence.
+withServerStreamCollectV2 ::
+    ( HasMethod CS.Queries m,
+      MonadIO n,
+      i ~ MethodInput CS.Queries m,
+      o ~ Seq (MethodOutput CS.Queries m) ) =>
+    -- |The procedure to invoke.
+    RPC CS.Queries m ->
+    -- |The procedure input.
+    i ->
+    -- |A mapping of the collected result.
+    (GRPCResult o -> b) ->
+    ClientMonad n b
+withServerStreamCollectV2 method input =
+    withServerStreamCallbackV2 method input mempty handler
+    where handler acc o = return $ acc <> pure o
+
+-- |Invoke a streaming GRPC API endpoint and return the (collected) results.
+-- Takes a `fold`-like callback and an accumulator, and returns the result
+-- of folding through each stream element, once the stream terminates.
+withServerStreamCallbackV2 ::
+    ( HasMethod CS.Queries m,
+      MonadIO n,
+      i ~ MethodInput CS.Queries m,
+      o ~ MethodOutput CS.Queries m ) =>
+    -- |The procedure to invoke.
+    RPC CS.Queries m ->
+    -- |The procedure input.
+    i ->
+    -- |An initial `fold`-like accumulator that
+    -- is updated with the handler each time a
+    -- stream element arrives.
+    a ->
+    -- |A `fold`-like handler which is used
+    -- to process a stream object and update
+    -- the accumulator.
+    (a -> o -> ClientIO a) ->
+    -- |A mapping of the accumulated result.
+    (GRPCResult a -> b) ->
+    ClientMonad n b
+withServerStreamCallbackV2 method input acc handler k =
+    withGRPCCoreV2 callHelper k
+    where
+        -- This is simply a handler which conforms to the one required by
+        -- `rawStreamServer`. This is here, so we may ignore the response
+        -- headers in the handler, just for the sake of ergonomics.
+        handler' acc' _hds streamObj = handler acc' streamObj
+        -- Helper that invokes the streaming call with the above callback.
+        -- This is here so we may leverage `withGRPCCoreV2`.
```

Comment:
The thing is, withGRPCCoreV2 is a bit tricky to use for streaming calls.

It is fine for calls that will immediately collect everything. But not for things like getBlocks and getFinalizedBlocks.

The reason being that it uses a RW lock internally. Meaning that during that stream you will not be able to handle any failures and reconnect.

In the V1 it made sense since queries were all unary. And it also makes sense I suppose for streaming where you collect immediately.